component/common/loki/client/internal/marker_file_handler.go

```diff
@@ -0,0 +1,96 @@
+package internal
+
+import (
+	"fmt"
+	"github.com/grafana/agent/component/common/loki/wal"
+	"os"
+	"path/filepath"
+	"strconv"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/prometheus/prometheus/tsdb/fileutil"
+)
+
+const (
+	MarkerFolderName = "remote"
+	MarkerFileName   = "segment"
+)
+
+// MarkerFileHandler is a file-backed wal.Marker, that also allows one to write to the backing store as particular
+// segment number as the last one marked.
+type MarkerFileHandler interface {
+	wal.Marker
+
+	// MarkSegment writes in the backing file-store that a particular segment is the last one marked.
+	MarkSegment(segment int)
+}
+
+type markerFileHandler struct {
+	logger                    log.Logger
+	lastMarkedSegmentFilePath string
+}
+
+var (
+	_ MarkerFileHandler = (*markerFileHandler)(nil)
+)
+
+// NewMarkerFileHandler creates a new markerFileHandler.
+func NewMarkerFileHandler(logger log.Logger, walDir string) (MarkerFileHandler, error) {
+	markerDir := filepath.Join(walDir, MarkerFolderName)
+	// attempt to create dir if doesn't exist
+	if err := os.MkdirAll(markerDir, 0o777); err != nil {
```

Comment:
777 feels a bit excessive.

Reply:
what do you think should be a sensible perm for the folder?

Reply:
I tend to go with 600/666, giving arbitrary files execute makes me shiver a bit.

Reply:
Let's go with 600, which is the most restrictive